File: test-suite-kotlin/src/test/kotlin/io/micronaut/views/docs/fieldset/AuthorFetcher.kt
package io.micronaut.views.docs.fieldset

import io.micronaut.views.fields.elements.Option
import io.micronaut.views.fields.fetchers.OptionFetcher
import io.micronaut.views.fields.messages.Message
import jakarta.inject.Singleton

//tag::clazz[]
@Singleton
class AuthorFetcher(private val authorRepository: AuthorRepository) : OptionFetcher<Long> {

    override fun generate(type: Class<Long>): List<Option> {
        return authorRepository.findAll()
            .map { author ->
                Option.builder()
                    .value(author.id.toString())
                    .label(Message.of(author.title))
                    .build()
            }
    }

    override fun generate(instance: Long): List<Option> {
        return authorRepository.findAll()
            .map { author ->
                Option.builder()
                    .selected(author.id == instance)
                    .value(author.id.toString())
                    .label(Message.of(author.title))
                    .build()
            }
    }
}
//end::clazz[]

File: test-suite-python/src/test/python/micronaut/views/docs/fieldset/AuthorFetcher.py
from jakarta.inject import Singleton
from micronaut.views.fields.elements import Option
from micronaut.views.fields.fetchers import OptionFetcher
from micronaut.views.fields.messages import Message

from .AuthorRepository import AuthorRepository


# tag::clazz[]
@Singleton
class AuthorFetcher(OptionFetcher[int]):

    def __init__(self, author_repository: AuthorRepository):
        self.author_repository = author_repository

    def generate(self, type_or_instance) -> list[Option]:
        options = []
        for author in self.author_repository.find_all():
            builder = (Option.builder()
                       .value(str(author.id))
                       .label(Message.of(author.title)))
            if isinstance(type_or_instance, int):
                builder.selected(author.id == type_or_instance)
            options.append(builder.build())
        return options
# end::clazz[]
